interpreter/cling/lib/Interpreter/Interpreter.cpp

```diff
@@ -666,6 +695,62 @@ namespace cling {
   }
 
   Interpreter::CompilationResult
+  Interpreter::codeComplete(const std::string& line, size_t& cursor,
+                            std::vector<std::string>& completions) const {
+
+    const char * const argV = "cling";
+    std::string llvmdir = this->getCI()->getHeaderSearchOpts().ResourceDir;
+    std::string extra_part = "/lib/clang/3.9.0";
+    std::string::size_type i = llvmdir.find(extra_part);
+    if (i != std::string::npos)
+      llvmdir.erase(i, extra_part.length());
+
+    cling::Interpreter childInterpreter(*this, 1, &argV, llvmdir.c_str());
+
+    auto childCI = childInterpreter.getCI();
+    clang::Sema &childSemaRef = childCI->getSema();
+
+    // Create the CodeCompleteConsumer with InterpreterCallbacks
+    // from the parent interpreter and set the consumer for the child
+    // interpreter.
+    ClingCodeCompleteConsumer* consumer = new ClingCodeCompleteConsumer(
+              this->getCI()->getFrontendOpts().CodeCompleteOpts, llvm::errs(),
+              completions);
+    // Child interpreter CI will own consumer!
+    childCI->setCodeCompletionConsumer(consumer);
+    childSemaRef.CodeCompleter = consumer;
+
+    // Ignore diagnostics when we tab complete.
+    // This is because we get redefinition errors due to the import of the decls.
+    clang::IgnoringDiagConsumer* ignoringDiagConsumer =
+                                            new clang::IgnoringDiagConsumer();
+    childSemaRef.getDiagnostics().setClient(ignoringDiagConsumer, true);
+
+    auto ownerDiagConsumer =
+                        this->getCI()->getSema().getDiagnostics().takeClient();
```

Comment:
You seem to be using `this->getCI()->getSema().getDiagnostics()` a lot - can you create a local variable for it?